// @solana-developers/helpers/dist/esm/lib/transaction.js
import { ComputeBudgetProgram, PublicKey, TransactionMessage, VersionedTransaction } from "@solana/web3.js";
import { getErrorFromRPCResponse } from "./logs.js";
export const confirmTransaction = async (connection, signature, commitment = "finalized") => {
    const block = await connection.getLatestBlockhash();
    const rpcResponse = await connection.confirmTransaction({
        signature,
        ...block,
    }, commitment);
    getErrorFromRPCResponse(rpcResponse);
    return signature;
};
// Was getSimulationUnits
// Credit https://twitter.com/stegabob, originally from
// https://x.com/stegaBOB/status/1766662289392889920
export const getSimulationComputeUnits = async (connection, instructions, payer, lookupTables) => {
    const testInstructions = [
        // Set an arbitrarily high number in simulation
        // so we can be sure the transaction will succeed
        // and get the real compute units used
        ComputeBudgetProgram.setComputeUnitLimit({ units: 1400000 }),
        ...instructions,
    ];
    const testTransaction = new VersionedTransaction(new TransactionMessage({
        instructions: testInstructions,
        payerKey: payer,
        // RecentBlockhash can by any public key during simulation
        // since 'replaceRecentBlockhash' is set to 'true' below
        recentBlockhash: PublicKey.default.toString(),
    }).compileToV0Message(lookupTables));
    const rpcResponse = await connection.simulateTransaction(testTransaction, {
        replaceRecentBlockhash: true,
        sigVerify: false,
    });
    getErrorFromRPCResponse(rpcResponse);
    return rpcResponse.value.unitsConsumed || null;
};
